### PR TITLE
 Build sources before pushing to central

### PR DESCRIPTION
**Pull request by NatashaWso2**
assume yes to all queries and do not prompt confirmation when pushing to central.

The flow will be as follows:

- If the user has specified the flag `ballerina push <package> --y`, then a confirmation will not be prompted to the user before pushing and by default it will push the package to central

- If the user has not specified the flag , then we prompt the following confirmation:
"Push the artifact <org-name>/<package>:<version> to central  [yes/y, no/n]: (y) "
By default we push the package to central, if the user says no/n the package is not pushed.

This also contains some refactoring to the integration test cases of push, pull, search and install with these new changes. 

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/9824

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InstallCommand.java [hunk 1]
```diff
@@ -50,6 +50,9 @@
     @Parameter(names = {"--sourceroot"}, description = "path to the directory containing source files and packages")
     private String sourceRoot;
 
+    @Parameter(names = {"--build"}, description = "build on push")
```
Comment (hemikak) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InstallCommand.java]: Shall we make it more descriptive ?

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InstallCommand.java]: Added proper descriptions with 5cbe69b

File: tests/ballerina-tools-integration-test/src/test/resources/packaging/Package.md [hunk 2]
```diff
@@ -1,37 +0,0 @@
-# Test Connector
```
Comment (Kishanthan) [inline tests/ballerina-tools-integration-test/src/test/resources/packaging/Package.md]: Why is this removed?

File: tests/ballerina-tools-integration-test/src/test/resources/packaging/functions.bal [hunk 3]
```diff
@@ -1,13 +0,0 @@
-import ballerina/io;
```
Comment (Kishanthan) [inline tests/ballerina-tools-integration-test/src/test/resources/packaging/functions.bal]: Have we replaced this with some other file?

File: tests/ballerina-tools-integration-test/src/test/java/org/ballerinalang/test/packaging/PackagingTestsUtils.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Util methods needed for Packaging test cases.
+ */
```
Comment (Kishanthan) [inline tests/ballerina-tools-integration-test/src/test/java/org/ballerinalang/test/packaging/PackagingTestsUtils.java]: Need version

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java [hunk 5]
```diff
@@ -75,8 +76,10 @@
      * @param packageName   path of the package folder to be pushed
      * @param sourceRoot    path to the directory containing source files and packages
      * @param installToRepo if it should be pushed to central or home
+     * @param pushToRepo true if confirmation is not required before push, else false
      */
-    public static void pushPackages(String packageName, String sourceRoot, String installToRepo) {
+    public static void pushPackages(String packageName, String sourceRoot, String installToRepo,
+                                    boolean pushToRepo) {
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: Isn't it pushToCentral ?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java [hunk 6]
```diff
@@ -98,11 +101,19 @@ public static void pushPackages(String packageName, String sourceRoot, String in
...
+        // Build the package
+        BuilderUtils.compileWithTestsAndWrite(prjDirPath, packageName, packageName, false, false, false, false);
 
         if (installToRepo == null) {
+            if (!pushToRepo) {
+                outStream.print("Push the artifact " + orgName + "/" + packageName + ":" + version  + " to central " +
+                                        "[yes/y, no/n]: (y) ");
+                Scanner scanner = new Scanner(System.in, Charset.defaultCharset().name());
+                String buildOnPush = scanner.nextLine().trim();
+                if (buildOnPush.equalsIgnoreCase("no") || buildOnPush.equalsIgnoreCase("n")) {
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: If some one say anything other than no/n , this will get installed right? If yes, shall we check for the yes/y and proceed?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java [hunk 7]
```diff
@@ -134,6 +145,15 @@ public static void pushPackages(String packageName, String sourceRoot, String in
             if (!installToRepo.equals("home")) {
                 throw new BLangCompilerException("Unknown repository provided to push the package");
             }
+            if (!pushToRepo) {
+                outStream.print("Install the artifact " + orgName + "/" + packageName + ":" + version  + " to the " +
+                                        "home repository [yes/y, no/n]: (y) ");
+                Scanner scanner = new Scanner(System.in, Charset.defaultCharset().name());
+                String buildOnPush = scanner.nextLine().trim();
+                if (buildOnPush.equalsIgnoreCase("no") || buildOnPush.equalsIgnoreCase("n")) {
```
Comment (Kishanthan) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: Same here. If some one say anything other than no/n , this will get installed right? If yes, shall we check for the yes/y and proceed?

File: tests/ballerina-tools-integration-test/src/test/resources/packaging/Package.md (hunk 2 above)
Comment (NatashaWso2) [reply tests/ballerina-tools-integration-test/src/test/resources/packaging/Package.md]: Previously before pushing we manually created the package zip to be pushed. But now we have used the init command to create the project and the build command to build the package zip. So we no longer need this.

File: tests/ballerina-tools-integration-test/src/test/resources/packaging/functions.bal (hunk 3 above)
Comment (NatashaWso2) [reply tests/ballerina-tools-integration-test/src/test/resources/packaging/functions.bal]: Previously before pushing we manually created the package zip to be pushed. But now we have used the init command to create the project and the build command to build the package zip. So we no longer need this file.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java (hunk 5 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: The same method is used both by the push command and the init command. In the case of push its "pushToCentral" and in the case of install "pushToHomeRepo". To be generic in both the cases I added it as "pushToRepo"

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java (hunk 7 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: Added the check with 3b98edc

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java (hunk 6 above)
Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/PushUtils.java]: Added the check with 3b98edc

File: tests/ballerina-tools-integration-test/src/test/java/org/ballerinalang/test/packaging/PackagingTestsUtils.java (hunk 4 above)
Comment (NatashaWso2) [reply tests/ballerina-tools-integration-test/src/test/java/org/ballerinalang/test/packaging/PackagingTestsUtils.java]: Added the version with 3b98edc

**Comment by Kishanthan** [comment]
Putting this on hold until all stakeholders agree on the approach. A discussion will be scheduled for this.

**Comment by Kishanthan** [comment]
Closing this until the discussion in concluded. Please reopen once we are finalized.
